src/libpmem2/map_posix.c

```diff
@@ -440,28 +429,39 @@ pmem2_map_new(struct pmem2_map **map_ptr, const struct pmem2_config *cfg,
 			goto err;
 		}
 
-		reserv = (char *)rsv_addr + rsv_offset;
-		if ((size_t)reserv % alignment) {
+		if (rsv_offset + reserved_length > rsv_size) {
+			ret = PMEM2_E_LENGTH_OUT_OF_RANGE;
+			ERR(
+				"Reservation %p has not enough space for the intended content",
+					rsv);
+			goto err;
+		}
+
+		reserv_region = (char *)rsv_addr + rsv_offset;
+		if ((size_t)reserv_region % alignment) {
 			ret = PMEM2_E_ADDRESS_UNALIGNED;
 			ERR(
 				"base mapping address %p (virtual memory reservation address + offset)" \
 				" is not a multiple of %zu required by device DAX",
-					reserv, alignment);
+					reserv_region, alignment);
 			goto err;
 		}
 
-		reserved_length = roundup(content_length, Pagesize);
-
-		map->addr = reserv;
-		map->content_length = content_length;
+		/* lock the mapping, until it's finalized */
+		util_rwlock_wrlock(&vm_rsv_lock);
 
-		/* register wanted vm reservation region */
-		ret = vm_reservation_map_register(cfg->reserv, map);
-		if (ret)
-			goto err;
+		/* check if the region in the reservation is occupied */
+		if (vm_reservation_map_find(rsv, rsv_offset, reserved_length)) {
```

Comment:
This function should be called something like `vm_reservation_map_find_acquire()` or something. It would make sure that no other thread can get the same reservation.

Reply:
Done.